core/net/rime/collect-neighbor.c

```diff
@@ -76,241 +71,187 @@ MEMB(collect_neighbors_mem, struct collect_neighbor, MAX_COLLECT_NEIGHBORS);
+         * the new one.
+         */
+        return 0;
       }
     }
 
-    /* Only add this new neighbor if its rtmetric is lower than the
-       one it would replace. */
-    if(nrtmetric < worst_rtmetric) {
-      n = worst_neighbor;
-    }
+    /* Now, we should have space to add the new neighbor */
+    n = nbr_table_add_lladdr(
+      collect_nbr_table,
+      addr,
+      NBR_TABLE_REASON_UNDEFINED,
+      NULL);
+
     if(n != NULL) {
-      PRINTF("collect_neighbor_add: not on list, not allocated, recycling %d.%d\n",
-             n->addr.u8[0], n->addr.u8[1]);
+      n->age = 0;
+      n->rtmetric = nrtmetric;
+      collect_link_estimate_new(&n->le);
+      n->le_age = 0;
+      return 1;
     }
   }
-
-  if(n != NULL) {
-    n->age = 0;
-    linkaddr_copy(&n->addr, addr);
-    n->rtmetric = nrtmetric;
-    collect_link_estimate_new(&n->le);
-    n->le_age = 0;
-    return 1;
-  }
   return 0;
 }
 /*---------------------------------------------------------------------------*/
-list_t
-collect_neighbor_list(struct collect_neighbor_list *neighbors_list)
+void
+collect_nbr_rm(collect_nbr_t *n)
 {
-  if(neighbors_list == NULL) {
-    return NULL;
-  }
-
-  return neighbors_list->list;
+  nbr_table_remove(collect_nbr_table, n);
 }
 /*---------------------------------------------------------------------------*/
-void
-collect_neighbor_list_remove(struct collect_neighbor_list *neighbors_list,
-                             const linkaddr_t *addr)
+collect_nbr_t *
+collect_nbr_best(void)
 {
-  struct collect_neighbor *n;
-
-  if(neighbors_list == NULL) {
-    return;
+  collect_nbr_t *n;
+  collect_nbr_t *best_nbr = NULL;
+  uint16_t best_rtmetric = RTMETRIC_MAX;
+
+  for(n = nbr_table_head(collect_nbr_table);
+      n != NULL;
+      n = nbr_table_next(collect_nbr_table, n)) {
+    if(n->rtmetric < best_rtmetric) {
+      best_nbr = n;
+      best_rtmetric = n->rtmetric;
+    }
   }
 
-  n = collect_neighbor_list_find(neighbors_list, addr);
-
-  if(n != NULL) {
-    list_remove(neighbors_list->list, n);
-    memb_free(&collect_neighbors_mem, n);
-  }
+  return best_nbr;
 }
 /*---------------------------------------------------------------------------*/
-struct collect_neighbor *
-collect_neighbor_list_best(struct collect_neighbor_list *neighbors_list)
+collect_nbr_t *
+collect_nbr_worst(void)
 {
-  struct collect_neighbor *n, *best;
-  uint16_t rtmetric;
-
-  rtmetric = RTMETRIC_MAX;
-  best = NULL;
-
-  if(neighbors_list == NULL) {
-    return NULL;
-  }
-
-  /*  PRINTF("%d: ", node_id);*/
-  PRINTF("collect_neighbor_best: ");
-
-  /* Find the neighbor with the lowest rtmetric + linkt estimate. */
-  for(n = list_head(neighbors_list->list); n != NULL; n = list_item_next(n)) {
-    PRINTF("%d.%d %d+%d=%d, ",
-           n->addr.u8[0], n->addr.u8[1],
-           n->rtmetric, collect_neighbor_link_estimate(n),
-           collect_neighbor_rtmetric(n));
-    if(collect_neighbor_rtmetric_link_estimate(n) < rtmetric) {
```

Comment:
In the nbr_table version the best neighbor is only calculated based on the rtmetric not the rtmetric+link_estimate score. The rtmetric however is only updated for neighbors we hear, thus if a parent neighbor goes away or we moved we would still try to use it for its potentially better rtmetric even if we can't reach it anymore.